chore(docs): Update the Basic Architecture page (#58)
openebs/mayastor-docs: reference/basic-architecture.md

| Name | Resource Type | Function | Frequency in Cluster |
| :--- | :--- | :--- | :--- |
| **Control Plane** |
| agent-core | Deployment | Principle control plane actor | Single |
| csi-controller | Deployment | Hosts Mayastor's CSI controller implementation and CSI provisioner side car| Single |
| api-rest | Pod | Hosts the public API REST server | Single |

> | *api-rest*  | Service | Exposes the REST API server via NodePort |

Reviewer: @tiagolobocastro is there a requirement to expose this on nodePortl?

Reviewer 2: No, we don't need to expose any ports now

Author: Removed `via NodePort`

| operator-diskpool | Deployment | Hosts DiskPool operator | Single |
| csi-node| DaemonSet | Hosts CSI Driver node plugin containers | All worker nodes |
| etcd | StatefulSet | Hosts etcd Server container | Configurable(<i>Recommended: Three replicas)</i> |
| etcd | Service | Exposes etcd DB endpoint | Single |
| etcd-headless | Service | Exposes etcd DB endpoint | Single |
| io-engine| DaemonSet | Hosts Mayastor I/O engine| User-selected nodes |

> | DiskPool | CRD | Declares a DiskPool's desired state and reflects its current state | User-defined, one or many |

Reviewer: In the description, should we say Mayastor pool's desired state...

| **Additional components**  |
| metrics-exporter-pool | Sidecar container (within io-engine DaemonSet)| Exports pool related metrics in Prometheus format | All worker nodes |
| pool-metrics-exporter | Service| Exposes exporter API endpoint to Prometheus | Single |
| promtail | DaemonSet| Scrapes logs of Mayastor-specific pods and exports them to Loki| All worker nodes |
| loki | StatefulSet| Stores the historical logs exported by promtail pods | Single |
| loki | Service| Exposes the Loki API endpoint via ClusterIP | Single |

## Component Roles

### io-engine

> The io-engine pods encapsulate Mayastor containers, which implement the I/O path from the block devices at the persistence layer, up to the relevant initiators on the worker nodes mounting volume claims.

Reviewer: pod

> The instance of the Mayastor running inside the container performs four major classes of functions:

Reviewer: Missing continuity between the sentence at line-34 and this.

> The instance of the Mayastor running inside the container performs four major classes of functions:

Reviewer 3: ```suggestion
The Mayastor process running inside this container performs four major functions:
```
(It's a bit wordy)

> - Presents a gRPC interface to the components which allow it to orchestrate the creation, configuration, and deletion of Mayastor-managed objects hosted by that instance.

Reviewer: This can be the 4th point and be modified as "Provides a gRPC interface service to orchestrate the creation, deletion and management of the above objects, hosted on that node."

### csi-node

The csi-node pods within a cluster implement the node plugin component of Mayastor's CSI driver. As such, their function is to orchestrate the mounting of Mayastor-provisioned volumes on worker nodes on which application pods consuming those volumes are scheduled. By default, a csi-node pod is scheduled on every node in the target cluster, as determined by a DaemonSet resource of the same name. Each of these pods encapsulates two containers, csi-node, and csi-driver-registrar.

The node plugin does not need to run on every worker node within a cluster and this behavior can be modified, if desired, through the application of appropriate node labeling and the addition of a corresponding nodeSelector entry within the pod spec of the csi-node DaemonSet. However, it should be noted that if a node does not host a plugin pod, then it will not be possible to schedule an application pod on it, which is configured to mount Mayastor volumes.

### etcd
   
etcd is a distributed reliable key-value store for the critical data of a distributed system. Mayastor uses etcd as a reliable persistent store for its configuration and state data.

### Supportability:

> The supportability tool is used to create support bundles (archive files) by interacting with multiple services present in the system. These bundles contain information about the entire Mayastor system, as well as specific Mayastor resources like volumes, pools and nodes, and can be used for debugging. It can collect the following information:

Reviewer: "present in the cluster where Mayastor is installed".

The tool can collect the following information:

> - Topological information of Mayastor's resource(s) by interacting with the REST service

Reviewer: Mayastor resources like (mention the list)

- Historical logs by interacting with Loki. If Loki is unavailable, it interacts with the kube-apiserver to fetch logs.
- Mayastor-specific Kubernetes resources by interacting with the kube-apiserver
- Mayastor-specific information from etcd (internal) by interacting with the etcd server.

### Loki:

    [Loki](https://grafana.com/oss/loki/) aggregates and centrally stores logs from all Mayastor containers which are deployed in the cluster.

### Promtail:
     

>     Promtail is a log collector built specifically for Loki. It uses the configuration file for target discovery and includes analogous features for labeling, transforming, and filtering logs from containers before ingesting them to Loki.

Reviewer: For the sake of heirarchy, I think we should keep Loki and Promtail, and also kubectl supportability commands under the supportability section giving the notion of a supportability stack.

Reviewer 3: I think we should leave this small reference here - it's purpose to to just inform the user that this pod is an expected part of their deployment, and what it is doing on their cluster.  The actual "care and feeding" can indeed be part of the supportability section.

Reviewer: In that case, I notice couple of inconsistencies here:
1. There is a "supportability" section above this "Promtail" section, and there is no tangible entity representing "supportability" as in a deployment/daemonset.
2. We don't have sections for other components here like api-server, core, pool-operator, etc.

Author: Based on our call earlier, I will be addressing these as part of another PR.

>     Promtail is a log collector built specifically for Loki. It uses the configuration file for target discovery and includes analogous features for labeling, transforming, and filtering logs from containers before ingesting them to Loki.

Reviewer: The external link for Promtail can be provided.
